From b11851c2f3f07150a65365ef68ea4e95a0aeb05e Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:26:05 -0400
Subject: [tools] perf/evlist: Introduce perf_evlist__strerror_tp method

Message-id: <1408462094-14194-197-git-send-email-jolsa@redhat.com>
Patchwork-id: 88189
O-Subject: [PATCH RHEL7.1 BZ1131394 196/325] perf evlist: Introduce perf_evlist__strerror_tp method
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 6ef068cb8e77784431b6c80adc49d0b0a6a5df66
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Thu Oct 17 12:07:58 2013 -0300

description
===========
Out of 'perf trace', should be used by other tools that uses
tracepoints.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index 90582e3..bfde691 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -1816,27 +1816,11 @@ out:
  trace->live = false;
  return err;
 out_error_tp:
- switch(errno) {
- case ENOENT:
-  fputs("Error:\tUnable to find debugfs\n"
-        "Hint:\tWas your kernel was compiled with debugfs support?\n"
-        "Hint:\tIs the debugfs filesystem mounted?\n"
-        "Hint:\tTry 'sudo mount -t debugfs nodev /sys/kernel/debug'\n",
-        trace->output);
-  break;
- case EACCES:
-  fprintf(trace->output,
-   "Error:\tNo permissions to read %s/tracing/events/raw_syscalls\n"
-   "Hint:\tTry 'sudo mount -o remount,mode=755 %s'\n",
-   debugfs_mountpoint, debugfs_mountpoint);
-  break;
- default: {
-  char bf[256];
-  fprintf(trace->output, "Can't trace: %s\n",
-   strerror_r(errno, bf, sizeof(bf)));
- }
-  break;
- }
+{
+ char errbuf[BUFSIZ];
+ perf_evlist__strerror_tp(evlist, errno, errbuf, sizeof(errbuf));
+ fprintf(trace->output, "%s\n", errbuf);
+}
  goto out_delete_evlist;
 }
 
diff --git a/tools/perf/util/evlist.c b/tools/perf/util/evlist.c
index a7d7786..81a1b3a 100644
--- a/tools/perf/util/evlist.c
+++ b/tools/perf/util/evlist.c
@@ -1133,3 +1133,30 @@ size_t perf_evlist__fprintf(struct perf_evlist *evlist, FILE *fp)
 
  return printed + fprintf(fp, "\n");;
 }
+
+int perf_evlist__strerror_tp(struct perf_evlist *evlist __maybe_unused,
+        int err, char *buf, size_t size)
+{
+ char sbuf[128];
+
+ switch (err) {
+ case ENOENT:
+  scnprintf(buf, size, "%s",
+     "Error:\tUnable to find debugfs\n"
+     "Hint:\tWas your kernel was compiled with debugfs support?\n"
+     "Hint:\tIs the debugfs filesystem mounted?\n"
+     "Hint:\tTry 'sudo mount -t debugfs nodev /sys/kernel/debug'");
+  break;
+ case EACCES:
+  scnprintf(buf, size,
+     "Error:\tNo permissions to read %s/tracing/events/raw_syscalls\n"
+     "Hint:\tTry 'sudo mount -o remount,mode=755 %s'\n",
+     debugfs_mountpoint, debugfs_mountpoint);
+  break;
+ default:
+  scnprintf(buf, size, "%s", strerror_r(err, sbuf, sizeof(sbuf)));
+  break;
+ }
+
+ return 0;
+}
diff --git a/tools/perf/util/evlist.h b/tools/perf/util/evlist.h
index acdf2b6..8ffb117 100644
--- a/tools/perf/util/evlist.h
+++ b/tools/perf/util/evlist.h
@@ -170,6 +170,8 @@ static inline struct perf_evsel *perf_evlist__last(struct perf_evlist *evlist)
 
 size_t perf_evlist__fprintf(struct perf_evlist *evlist, FILE *fp);
 
+int perf_evlist__strerror_tp(struct perf_evlist *evlist, int err, char *buf, size_t size);
+
 static inline unsigned int perf_mmap__read_head(struct perf_mmap *mm)
 {
  struct perf_event_mmap_page *pc = mm->base;
-- 
1.7.1